saml/src/main/java/com/linecorp/armeria/server/saml/SamlService.java

```diff
@@ -0,0 +1,222 @@
+
+        final CompletionStage<AggregatedHttpMessage> f;
+        if (portConfigHolder.isDone()) {
+            f = req.aggregate();
+        } else {
+            f = portConfigHolder.future().thenCompose(unused -> req.aggregate());
+        }
+        return HttpResponse.from(f.handle((msg, cause) -> {
+            if (cause != null) {
+                return HttpResponse.of(HttpStatus.BAD_REQUEST);
+            }
+            final SamlPortConfig portConfig = portConfigHolder.config().get();
+            if (portConfig.scheme().isTls() != ctx.sessionProtocol().isTls()) {
+                return HttpResponse.of(HttpStatus.BAD_REQUEST);
+            }
+
+            // Use user-specified hostname if it exists.
+            // If there's no hostname set by a user, the default virtual hostname will be used.
+            final String defaultHostname = firstNonNull(sp.hostname(), ctx.virtualHost().defaultHostname());
+            return func.serve(ctx, msg, defaultHostname, portConfig);
+        }));
+    }
+
+    /**
+     * A wrapper class which holds parameters resolved from a query string.
+     */
+    static final class SamlParameters {
+        private final Map<String, List<String>> parameters;
+
+        /**
+         * Creates a {@link SamlParameters} instance with the specified {@link AggregatedHttpMessage}.
+         */
+        SamlParameters(AggregatedHttpMessage msg) {
+            requireNonNull(msg, "msg");
+            final MediaType contentType = msg.headers().contentType();
+
+            final QueryStringDecoder decoder;
+            if (contentType != null && contentType.belongsTo(MediaType.FORM_DATA)) {
+                final String query = msg.content().toString(
+                        contentType.charset().orElse(StandardCharsets.UTF_8));
+                decoder = new QueryStringDecoder(query, false);
+            } else {
+                final String path = msg.path();
+                assert path != null : "path";
+                decoder = new QueryStringDecoder(path, true);
+            }
+
+            parameters = decoder.parameters();
+        }
+
+        /**
+         * Returns the first value of the parameter with the specified {@code name}.
+         *
+         * @throws NullPointerException if the specified {@code name} is {@code null}
+         * @throws SamlException if a parameter with the specified {@code name} does not exist
+         */
+        String getFirstValue(String name) throws SamlException {
+            final String value = getFirstValueOrNull(name);
+            if (value == null) {
+                throw new SamlException("failed to get the value of a parameter: " + name);
+            }
+            return value;
+        }
+
+        /**
+         * Returns the first value of the parameter with the specified {@code name}. If it does not exist,
+         * {@code null} is returned.
+         *
+         * @throws NullPointerException if the specified {@code name} is {@code null}
```

Comment:
Perhaps, we can omit this statement like we did in other methods in which use `requireNonNull`.

Reply:
Ok, removed